### PR TITLE
Add Continuous Integration via GitHub Actions

### LINKED ISSUE #8: CI pipeline
We should use GitHub Actions to ensure that code continues to pass our quality requirements- pass tests, lint, etc. These actions should be set up to run against incoming pull requests.

### PR DESCRIPTION
**Pull request by Lucretiel**
This Pull Request adds a GitHub action that runs our unit tests which will run against all pull requests.

Fixes #8 

Additionally, a project maintainer should go into project settings and enable Branch Protection (see attached screenshot), to *require* that these tests pass before Pull Requests can be merged

image

Additionally, this PR:

- Adds a yarn lockfile, to ensure reliable / reproducible builds
- Fixes some tests on Windows

### CODE REVIEW

**Comment by darrellanderson** [comment]
DivineDuck (more familiar with GitHub action) is going to take a look at this soon.

**Comment by Lucretiel** [comment]
I'd also recommend requiring pull requests in general (also in the Branch Protection settings), just to help ensure all new code has at least the potential for review

**Comment by darrellanderson** [comment]
Working through pull requests instead of branches is a good idea.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  pull_request:
+    branches: [main]
+
+jobs:
+  ci:
+    strategy:
+      fail-fast: true
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+        node-version: [16.x]
```
Comment (duckman312) [inline .github/workflows/CI.yml:18]: we should enforce a 16.x node version in the package to match. unless TTPG requires 16.x to use, then it's redundent

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+      # Set up Node and yarn; use cached dependencies
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v2
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: "yarn"
+
+      - run: yarn install --frozen-lockfile
+      - run: yarn generate:langfile
```
Comment (duckman312) [inline .github/workflows/CI.yml:36]: I'm a little confused. why does `generate:langfile` need to be run on PR action?

**Review by duckman312 — APPROVED** [review]
looks fine. I'll also make the config changes you mentioned

**Comment by duckman312** [comment]
actually i'll just promote you to contributor and you're welcome to make the changes you've specified. you clearly have a setup in mind and i don't want to get in the way

File: .github/workflows/CI.yml (hunk 2 above)
Comment (Lucretiel) [reply .github/workflows/CI.yml:36]: There are unit tests that depend on the existence of `langfile.js`, but the build steps that generate that file are dependent on the configuration created by `yarn setup` (which requires the local installation of Tabletop Playground and therefore isn't really relevant to CI / unit tests). I added `generate:langfile` to skip all of the source copying / symlinking and just generate the langfile.

**Comment by Lucretiel** [comment]
Thanks for the invite! It looks like though that (per https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule) only project administrators can modify branch protection rules. Totally understand if you don't want to give me that level of authority.

**Comment by duckman312** [comment]
Should be setup now. let me know if i missed anything

**Pull request merged**